### ISSUE TITLE
Inconsistent "user" ID handling with XrdHttp / XRootD

### ISSUE THREAD

**Issue by olifre** [xrootd/xrootd#1268]
This could also be a configuration issue, if so, please bear with me ;-). I'm using XRootD 5.0.0 in this test (a self-compiled version from master after the `http.tlsreuse` parameter was introduced). 

I'm trying to access the path  `/data/belle`, given the following configuration (slightly trimmed irrelevant parts):
```
acc.audit deny grant
acc.authdb /etc/xrootd/auth_file-grid
acc.authrefresh 60
all.export /data/belle r/w nostage
all.manager xrootd-dev.example.com:1213
all.role server
all.sitename MYSITE
cms.allow host xrootd-dev.example.com
cms.trace all -debug
http.cadir /etc/grid-security/certificates
http.desthttps yes
http.header2cgi Authorization authz
http.secxtractor /usr/lib64/libXrdVoms.so certfmt=raw|grpopt=useall|vos=atlas,belle,ops,dteam,wlcg|grps=/atlas,/atlas/de,/belle,/ops,/dteam,/wlcg|dbg
http.selfhttps2http no
http.tlsreuse on
if exec xrootd
xrd.protocol XrdHttp /usr/lib64/libXrdHttp.so
fi
sec.protocol /usr/lib64 gsi -dlgpxy:1 -exppxy:=creds -ca:1 -crl:3 -gridmap:/dev/null -cert:/etc/grid-security/hostcert.pem -key:/etc/grid-security/hostkey.pem -certdir:/etc/grid-security/certificates -vomsfun:/usr/lib64/libXrdVoms.so -vomsfunparms:certfmt=raw|grpopt=useall|vos=atlas,belle,ops,dteam,wlcg|grps=/atlas,/atlas/de,/belle,/ops,/dteam,/wlcg|dbg
xrd.port 1094
xrd.tls /etc/grid-security/hostcert.pem /etc/grid-security/hostkey.pem
xrd.tlsca certdir /etc/grid-security/certificates
xrd.trace conn
xrootd.chksum adler32 crc32 md5
xrootd.seclib /usr/lib64/libXrdSec.so
acc.spacechar _
```
with the following `auth_file-grid`:
```
= belleusr   o: belle g: /belle
= bellesadm  o: belle g: /belle u: HERES_THE_ID
x bellesadm  /data/belle a
x belleusr   /data/belle lr
```
I am trying with a VOMS proxy with:
* FQAN `/belle/Role=NULL/Capability=NULL`
* certificate DN `/C=DE/O=GermanGrid/OU=UniBonn/CN=Oliver Freyermuth`
* DN hash `aeb776c8.0`

Read access works fine, my problem is granting write access to a single "user" (without using a more elaborate tool such as LCMAPS). 

I've tried various IDs for `HERES_THE_ID` with mixed success:

| Value of `HERES_THE_ID` | Write access works via HTTP | Write access works via XRootD |
|--------------------------------------|-----------------------------------------|-------------------------------------------|
| /C=DE/O=GermanGrid/OU=UniBonn/CN=Oliver_Freyermuth | denied | denied | 
| Oliver_Freyermuth | denied | denied |
| aeb776c8.0 | denied | works! |

So I have two issues:
* Why does the "DN" with escaped space (using `acc.spacechar`, see https://github.com/xrootd/xrootd/issues/712 ) not work? 
* Why does the DN hash only work for a connection from an XRootD client, but not for a connection from a HTTP client? 

Note that when creating a connection via `xrdcp`, I see this in the logs:
```
200804 12:49:14 1255818 XrootdXeq: olifre.1656:33@mytestnode pub IP46 login as aeb776c8.0
```
and when connecting via `curl` using the proxy cert, I see:
```
200804 12:50:26 1255800 XrootdBridge: eyermuth.3:29@mytestnode login as eyermuth
```
Is it expected that when connecting via HTTP, the connection will not be mapped to an ID containing the certificate DN hash, or did I misconfigure something?

**Comment by abh3** [comment]
Hi Oliver,

You are not doing anything wrong per se. You just happen to be the first 
one that is actually trying to use the DN for access control and exposed 
an inconsistency that has long been forgotten. I say first because this 
issue has come up before a long time ago and people have since abondoned 
using the DN without, as you point out, also using either gridmap or 
LCMAPS to get a resonable "userid". I will look to see where the 
inconsistency comes from and let you know what can be done about it.

Andy

**Comment by olifre** [comment]
Hi Andy,

thanks for confirming! 
In our case, it's not urgent — it is just the most straightforward solution to our problem (Belle does not have as much DDM as the large experiments yet, so manual copies are needed, XRootD TPC is a perfect match, but we want to grant write permissions only to us site admins — and for essentially two DNs, LCMAPS and gridmap are a bit of overkill). So that sounds like a good chance to hunt down this inconsistency :-). 
An additional question, if I may: Is it expected that only the DN hash is accepted, and the full DN is not? Thinking about it, there's probably no clean way to support both, though. 

Cheers and thanks for looking into it,
Oliver

**Comment by xrootd-dev** [comment]
Hi Oliver,

Indeed, only one or the other (DN or hash) is supported with the given 
code, though nothing stops you from writing a plugin that does both :-)

Andy

**Comment by abh3** [comment]
So, given the understanding here, can we close the ticket?

**Comment by abh3** [comment]
I'll answer this in dribs and drabs as I figure out the issues so as to be speedy.

a) The reason that spacechar was not applied is that we don't apply it for "=" (anyuser) or '*" (alluser). I can't recall what the reasoning behind that restriction. So, please think about it and how those two rules are used and let me know your thoughts on whether we should lift the restriction. However, before you do that, consider point (b).

b) The dn hash is a specific feature of gsi authentication. Given that http does not use gsi, it never develops a hash. It's not clear whether we should change that or not. I suppose we could make it an option. Thoughts?

Now as for using the dn. It's problematic because as far as I can tell the order of the elements in the dn field is not standardized. Additionally, some of these are optional. So, doing a simple comparison is not sufficient. You have to parse the DN and compare element by element. That's something we don't do which makes the dn field not particularly useful. That also implies that the dn hash is also subject to this problem (though I will have to verify that). Generally, people have opted to map to a name using whatever algorithm is appropriate for the experiment. Consequently, there is no single way to codify the use of the dn field.

**Comment by olifre** [comment]
@abh3 Many thanks for this nice summary! 

> b) The dn hash is a specific feature of gsi authentication. Given that http does not use gsi, it never develops a hash. It's not clear whether we should change that or not. I suppose we could make it an option. Thoughts?

I think that if the DN can be used, consistent use of the hash is not as important anymore. My main reason to even consider using the hash was that using the DN failed due to the spacechar issue. 

Concerning the DN usage, I agree you are right: For any elaborate, experiment-specific mapping, an external algorithm is used and this will likely stay this way, especially for the large experiments. The use cases I am thinking of are related to smaller user bases, so some examples could be:
* Granting special privileges to single users (e.g. site admins, robots, debugging). I'm thinking of use cases for which the list is short (<10 people or robots) and manageable, e.g. small local user communities, site admins or even a DAQ robot uploading data. 
* Accepting only users from a specific OU of the OUs CA. 
* Blocking single users / DNs (e.g. "account hacked") before the official CRL is updated. 

The initial use case I had in mind when creating this issue was indeed just "granting users ingesting data write privileges", allowing regular users (granted permissions by VOMS extensions) to only read the data. But we also have other of the use cases listed above for which this might be helpful, and I think these are mostly generic so I believe they could be helpful to many users. 

Now for the implementation. Of course, you are right when you write:

> Now as for using the dn. It's problematic because as far as I can tell the order of the elements in the dn field is not standardized. Additionally, some of these are optional. So, doing a simple comparison is not sufficient. You have to parse the DN and compare element by element. That's something we don't do which makes the dn field not particularly useful.

However, since the external algorithms exist for the more complex use cases, and likely won't ever be fully replaced by anything thought up here, I wonder if the following approach would work: Match the string in the auth file against the DN, but allow for wildcards / a regular expression. Since the CAs are trusted to not violate each others namespaces / DNs, and the DNs are hierarchical, this might be sufficient. To give some examples, it would allow to:
* Match against something like `^/C=DE/O=GermanGrid/OU=UniBonn/.*` (or any other `OU`) without actually parsing the DN. 
* Explicitly match against a single DN (to exclude it or grant it special privileges). 
* Depending on the CAs which are trusted (e.g. a small user group may even have their own CA), matching by a single component of the DN might even be a reasonable use case. 

In the end, a regexp / wildcard solution would delegate the definition of what is acceptable to the user, who has to specify which components of the DN must match a specific pattern. I presume it might also be the "easiest" implementation, and would be sufficient for the use cases I can think of. 

Of course, you are right that string matching is not fully correct, and parsing the DN into components and treating these would be much more versatile — but also much more complex. This [article](https://stackoverflow.com/a/39841564/9679188) has a nice summary on why the elements of DNs "commonly" are in hierarchical order (due to inspiration of `X.509` by directory services). That's also the case for all DNs I have encountered so far, and what the commonly trusted CAs (CAs issuing web certificates or CAs in use in WLCG) ensure. However, it is not in any standard, so there is no guarantee for that to be the case for "everything X.509". 

Having a regexp / wildcard solution would delegate the wanted "strictness" and the definition of what is "acceptable"  to the user writing the auth file. Since the user also defines which CAs to trust by defining the trust store, this might be a reasonable assumption to make. What do you think? 

Now finally, to the last point:

> a) The reason that spacechar was not applied is that we don't apply it for "=" (anyuser) or '*" (alluser). I can't recall the reasoning behind that restriction. So, please think about it and how those two rules are used and let me know your thoughts on whether we should lift the restriction. However, before you do that, consider point (b).

I also can not think of a reasoning behind that restriction. I'll ponder about it a bit more, but I currently can't imagine a situation in which it would "hurt" to lift the restriction,

**Comment by abh3** [comment]
@olifre It's been two years of pondering. Have you gotten a workaround or de we still see what can be done about this inconsistency?

**Comment by olifre** [comment]
@abh3 Thanks for coming back to this! In fact, I have not found a workaround, but the issue has not been "pressing" enough in the past two years so I have used the XRootD protocol to access in case the individual certs had to be used. 

Going back to your original questions, I think:
* The restriction not to apply `spacechar` to `=` or `*` can be lifted — at least I don't see any way in which this would break existing use cases. 
* If that is in place, there's no real need to support the hash in HTTP. When using the DN directly, I think that both HTTP and XRootD should work the same, correct? 
* To support real-life use cases, though, a regexp matching would be required. 

What do you think?

**Comment by amadio** [comment]
@olifre Could you please test this again now that the id has been aligned between XRootD and HTTP protocols in

**Comment by olifre** [comment]
@amadio Sure, it will take me a bit, but I can test, of course. 
The commit you linked is about the monitoring ID being adapted, though — is it expected that this would really also affect the ID used for access control / authorization?

**Comment by amadio** [comment]
@olifre Indeed, I saw HTTP vs XRootD differences and thought about this issue, but it does seem not to be related.

**Comment by abh3** [comment]
Yes, @olifre the identifier is used for TPC rendezvous in EOS. So, making t different from what the client reported would make the TPC inoperable. The change simply made sure it was only affecting the monitoring system and nothing more. sine HTTP does it's own generation of the IS we don't really know what that affect is but I suspect it affects nothing as HTTP TPC is completely different.
